### PR TITLE
Rework entrypoint script (now script*s*)

### PR DESCRIPTION
**Pull request by JayH5**
* Add a separate entrypoint script for Celery so Celery commands can be run under the correct user and with limited processes easily.
* Rework `django-entrypoint.sh` to actually be an entrypoint script.
* Deprecate `APP_MODULE` env var. Set the app module for Gunicorn via `CMD` now.
* README work for all this.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -104,22 +136,21 @@ Note that by default Celery runs as many worker processes as there are processor
 * Celery option: `-c`/`--concurrency`
 
 #### `CELERY_BEAT`:
-Set this option to any non-empty value to have a [Celery beat](http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html) scheduler process run as well.
+Set this option to any non-empty value (e.g. "`1`") to have a [Celery beat](http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html) scheduler process run as well.
```
Comment (jerith) [inline README.md]: Do we need both kinds of quotes around the '1'?

File: celery-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+set -e
+
+_is_celery_command () {
+  local cmd="$1"; shift
+
+  python - <<EOF
+import sys
+from celery.bin.celery import CeleryCommand
+sys.exit(0 if '$cmd' in CeleryCommand.commands else 1)
+EOF
```
Comment (jerith) [inline celery-entrypoint.sh:11]: 😿

File: celery-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    set -- celery "$@"
+  fi
+fi
+
+if [ "$1" = 'celery' ]; then
+  # Set some common options if the env vars are set
+  set -- "$@" \
+      ${CELERY_APP:+--app "$CELERY_APP"} \
+      ${CELERY_BROKER:+--broker "$CELERY_BROKER"} \
+      ${CELERY_LOGLEVEL:+--loglevel "$CELERY_LOGLEVEL"}
```
Comment (jerith) [inline celery-entrypoint.sh:26]: What happens if these parameters have already been passed in? (Also for the next two blocks.)

File: django-entrypoint.sh [hunk 4]
```diff
@@ -1,63 +1,69 @@
 #!/usr/bin/env sh
 set -e
 
-if [ -z "$APP_MODULE" ]; then
-  echo "The \$APP_MODULE environment variable must be set to the WSGI application module name"
-  exit 1
+# No args or looks like options or the APP_MODULE for Gunicorn
+if [ "$#" = 0 ] || \
+    [ "${1#-}" != "$1" ] || \
+    echo "$1" | grep -Eq '^([_A-Za-z]\w*\.)*[_A-Za-z]\w*:[_A-Za-z]\w*$'; then
```
Comment (jerith) [inline django-entrypoint.sh:7]: 😿 😿 😿

File: django-entrypoint.sh [hunk 5]
```diff
@@ -1,63 +1,69 @@
...
+  # but here we do the minimum.
+  su-exec gunicorn django-admin migrate --noinput
 
-if [ -n "$SUPERUSER_PASSWORD" ]; then
-  echo "from django.contrib.auth.models import User
+  if [ -n "$SUPERUSER_PASSWORD" ]; then
+    echo "from django.contrib.auth.models import User
 if not User.objects.filter(username='admin').exists():
     User.objects.create_superuser('admin', 'admin@example.com', '$SUPERUSER_PASSWORD')
 " | su-exec gunicorn django-admin shell
```
Comment (jerith) [inline django-entrypoint.sh:33]: Should this not also be a HEREDOC? I don't really trust multiline strings.

**Review by jerith — COMMENTED** [review]
Is it worth factoring out a `gunicorn-entrypoint.sh` script as well? It looks like quite a lot of the `django-entrypoint.sh` code is in the `if [ "$1" = 'gunicorn' ]` block.

File: celery-entrypoint.sh (hunk 3 above)
Comment (JayH5) [reply celery-entrypoint.sh:26]: Bad things :'( It's basically this problem: https://github.com/praekeltfoundation/pydexec/pull/16 and I don't know how to solve it.

In the next PR I'm hoping to remove these extra options for Celery (or at least deprecate them). Celery itself will pick up the `CELERY_APP` env var, and the others should probably rather be set in the Django settings.

File: django-entrypoint.sh (hunk 4 above)
Comment (JayH5) [reply django-entrypoint.sh:7]: I almost ran some more Python code to see if this described a module that could be imported 😿

File: django-entrypoint.sh (hunk 5 above)
Comment (JayH5) [reply django-entrypoint.sh:33]: My thought was the same.. and I tried but I couldn't get it to work. The `django-admin shell` command doesn't take `-` like `python` does.

**Comment by JayH5** [comment]
@jerith: My first attempt was actually to write a `gunicorn-entrypoint.sh` script. The issue is that we need to determine whether to run all the startup steps or not...which amounts to checking whether stuff looks like a Gunicorn command or not...which is half a Gunicorn entrypoint script already.

File: celery-entrypoint.sh (hunk 2 above)
Comment (JayH5) [reply celery-entrypoint.sh:11]: The list of commands is long and varies between Celery 3.1 and 4.0 unfortunately 😿

**Review by jerith — APPROVED** [review]
:+1:

**Pull request merged**